velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -17,56 +17,121 @@
+      writeProtocol_(std::move(writeProtocol)),
+      partitionChannels_(getPartitionChannels(insertTableHandle_)),
+      partitionIdGenerator_(
+          partitionChannels_.size() > 0
+              ? std::make_shared<PartitionIdGenerator>(
+                    inputType_,
+                    partitionChannels_,
+                    HiveConfig::maxPartitionsPerWriters(
+                        connectorQueryCtx_->config()))
+              : nullptr) {
   VELOX_CHECK_NOT_NULL(
       writeProtocol_, "Write protocol could not be nullptr for HiveDataSink.");
 }
 
 std::shared_ptr<ConnectorCommitInfo> HiveDataSink::getConnectorCommitInfo()
     const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
+  std::vector<std::shared_ptr<const HiveWriterParameters>> toCommit;
+  toCommit.reserve(writerParameters_.size());
+
+  for (auto i = 0; i < writerParameters_.size(); i++) {
+    if (writerParameters_[i] != nullptr) {
+      toCommit.push_back(writerParameters_[i]);
+    }
+  }
+  return std::make_shared<HiveConnectorCommitInfo>(toCommit);
 }
 
 void HiveDataSink::appendData(VectorPtr input) {
-  // For the time being the hive data sink supports one file
-  // To extend it we can create a new writer for every
-  // partition
-  if (writers_.empty()) {
-    auto hiveWriterParameters =
-        std::dynamic_pointer_cast<const HiveWriterParameters>(
-            writeProtocol_->getWriterParameters(
-                insertTableHandle_,
-                connectorQueryCtx_,
-                std::make_shared<HiveConnectorWriteInfo>()));
-    VELOX_CHECK_NOT_NULL(
-        hiveWriterParameters,
-        "Hive data sink expects write parameters for Hive.");
-    writerParameters_.emplace_back(hiveWriterParameters);
-    writers_.emplace_back(createWriter(writerParameters_[0]));
+  input_ = std::dynamic_pointer_cast<RowVector>(input);
+
+  // Writing to unpartitioned table
+  if (partitionChannels_.size() == 0) {
+    if (writers_.empty()) {
+      writers_.resize(1);
+      writerParameters_.resize(1);
+      createWriterForPartition(0);
+    }
+    writers_[0]->write(input_);
+    return;
+  }
+
+  // Writing to partitioned table
+  partitionIdGenerator_->run(input_, partitionIds_);
+
+  uint64_t maxPartitionId = partitionIdGenerator_->getMaxPartitionId();
+
+  computePartitionRowCountsAndIndices(
+      maxPartitionId + 1, connectorQueryCtx_->memoryPool());
+
+  for (column_index_t i = 0; i < input_->childrenSize(); i++) {
+    input_->childAt(i)->loadedVector();
+  }
+
+  if (writers_.size() < (maxPartitionId + 1)) {
+    writers_.resize(maxPartitionId + 1);
+    writerParameters_.resize(maxPartitionId + 1);
+  }
+  for (vector_size_t partitionId = 0; partitionId <= maxPartitionId;
+       partitionId++) {
+    if (partitionSizes_[partitionId] == 0) {
+      continue;
+    }
+
+    if (!writers_[partitionId]) {
+      createWriterForPartition(partitionId);
+    }
+
+    RowVectorPtr writerInput = exec::wrap(
```

Comment:
Writing a single partition is a common use case. Consider optimizing for that (can be a follow-up PR).